# Momento evaluativo 3

Seminario <br>
Especialización en Analítica y Ciencia de Datos <br>
Universidad de Antioquia <br>
Noviembre 2023 <br>

Julieth Tatiana García Zuluaga - Grupo 12

# Decripción del dataset

El siguiente dataset es una extracción de la base de datos del foro de discusión de una empresa. Sus campos se seleccionaron de acuerdo al objetivo de realizar una segmentación de los usuarios de la aplicación según su comportamiento en el foro.

| Campo | Descripción | 
| :--- | :--- |
| username | Nombre de usuario en la aplicación |
| trust_level | Nivel de confianza del usuario en la aplicación |
| topics_entered | Temas ingresados por el usuario en el foro |  
| time_read | Tiempo de lectura del usuario en el foro |
| days_visited | Cantidad de días que ha visitado el foro |
| post_read_count | Cantidad de posts leídos en el foro |
| likes_given | Me gusta dados por el usuario en el foro |
| likes_received | Me gusta recibidos por el usuario en el foro |
| new_since | Fecha de creación del usuario en el foro |
| post_count | Cantidad de post realizados por el usuario |
| topic_count | Cantidad de temas realizas por el usuario |
| user_custom_field | Campo personalizado por el usuario en el foro |
| like_count | Cantidad de me gusta en los post del usuario en el foro |
| score | Puntaje del usuario por su participación en el foro |
| reads | Número de lecturas que ha recibido el usuario en el foro |
| word_count | Cantidad de palabras que contiene el post realizado por el usuario en el foro |
| image_upload_id | ID de la imagen insertada (Si aplica) |
| topic | Tema del post realizado por el usuario en el foro |
| category | Categoria del tema posteado por el usuario en el foro |
| category_description | Descripción de la categoría del temas posteado por el usuario en el foro |
| tag | Etiqueta otorgada al post por el sistema en el foro |
| given_date_likes | Fecha del me gusta otorgadao |
| post | Descripción del post publicado por el usuario en el foro |
| archetype | Arquetipo del tema |

Resumen de las características:
- Instancias: 2.949
- Columnas: 24
- Objetivo: Segmentación o clusterización de usuarios
***


# Objetivo

Segmentar y clasificar a los usuarios que utilizan la herramienta de foro de discusión en la Vicepresidencia de Tecnología de la empresa.

# Librerías

In [38]:
# Tratamiento de datos
import pandas as pd
import numpy as np

# Configuración de warnings
import warnings
warnings.filterwarnings('ignore')

# Cargar dataset

In [39]:
# Vista preliminar del data set
data = pd.read_csv('https://raw.githubusercontent.com/TatianaGarcia1128/foro_discusion_udea/main/dataset_users.csv')
data.head()

username  trust_level  topics_entered  time_read  days_visited  \
0   yvarela            4             520      81584           168   
1  ddurango            4              59       8999            65   
2  ddurango            4              59       8999            65   
3  ddurango            4              59       8999            65   
4   yvarela            4             520      81584           168   

   posts_read_count  likes_given  likes_received                   new_since  \
0              1259           71              61  2022-07-19 15:51:12.780824   
1               119            9              12  2022-07-19 13:45:13.649354   
2               119            9              12  2022-07-19 13:45:13.649354   
3               119            9              12  2022-07-19 13:45:13.649354   
4              1259           71              61  2022-07-19 15:51:12.780824   

   post_count  ...  reads word_count  image_upload_id  \
0          27  ...     94        107              NaN   
1          13  ...     44         87              NaN   
2          13  ...     44         87              NaN   
3          13  ...     44         87              NaN   
4          27  ...     94        107              NaN   

                              topic   category  \
0  Acerca de la categoría DevSecOps  DevSecOps   
1    Acerca de la categoría Galatea    Galatea   
2    Acerca de la categoría Galatea    Galatea   
3    Acerca de la categoría Galatea    Galatea   
4  Acerca de la categoría DevSecOps  DevSecOps   

                                category_description                    tag  \
0  Te invitamos a contribuir al proceso <strong>D...  continuous-compliance   
1                                                NaN         microfrontends   
2                                                NaN    distribution-router   
3                                                NaN                  figma   
4  Te invitamos a contribuir al proceso <strong>D...                angular   

   given_date                                               post archetype  
0  2022-10-28  Te invitamos a contribuir al proceso **DevSecO...   regular  
1  2022-09-29  (Sustituye este párrafo con una descripción co...   regular  
2  2022-07-25  (Sustituye este párrafo con una descripción co...   regular  
3  2022-09-16  (Sustituye este párrafo con una descripción co...   regular  
4  2022-07-27  Te invitamos a contribuir al proceso **DevSecO...   regular  

[5 rows x 24 columns]

In [40]:
# Revisar nulos, tipo de datos en cada variable y tamaño de la base de datos
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2949 entries, 0 to 2948
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   username              2949 non-null   object 
 1   trust_level           2949 non-null   int64  
 2   topics_entered        2949 non-null   int64  
 3   time_read             2949 non-null   int64  
 4   days_visited          2949 non-null   int64  
 5   posts_read_count      2949 non-null   int64  
 6   likes_given           2949 non-null   int64  
 7   likes_received        2949 non-null   int64  
 8   new_since             2949 non-null   object 
 9   post_count            2949 non-null   int64  
 10  topic_count           2949 non-null   int64  
 11  user_custom_field     2949 non-null   object 
 12  like_count            2949 non-null   int64  
 13  score                 2949 non-null   float64
 14  reads                 2949 non-null   int64  
 15  word_count           

El data set contiene 2949 instancias y 24 columnas, de las cuales 12 tienen valores tipo entero (int64), 2 de tipo decimal (float64) y las 10 restantes son tipo object. De acuerdo con la información mostrada, existen valores nulos en las variables 'image_upload_id' y 'category_description'. Además, la base de datos ocupa 553.1+ KB, y es fácil de procesar.

Ahora solo presenta valores nulos la variable 'category_description'.

In [41]:
# Estadística descriptiva de las variables númericas
data.describe()

trust_level  topics_entered     time_read  days_visited  \
count  2949.000000     2949.000000   2949.000000   2949.000000   
mean      3.886063      357.735504  55816.559512    130.492370   
std       0.463643      220.205834  34978.531139     51.127422   
min       2.000000       59.000000   5160.000000     43.000000   
25%       4.000000       59.000000   8999.000000     65.000000   
50%       4.000000      520.000000  81584.000000    168.000000   
75%       4.000000      520.000000  81584.000000    168.000000   
max       4.000000      520.000000  81584.000000    168.000000   

       posts_read_count  likes_given  likes_received   post_count  \
count       2949.000000  2949.000000     2949.000000  2949.000000   
mean         858.820956    48.721261       43.354018    21.616480   
std          543.090681    30.284533       23.995819     7.512162   
min          119.000000     1.000000        5.000000     5.000000   
25%          119.000000     9.000000       12.000000    13.000000   
50%         1259.000000    71.000000       61.000000    27.000000   
75%         1259.000000    71.000000       61.000000    27.000000   
max         1259.000000    71.000000       61.000000    27.000000   

       topic_count   like_count        score        reads   word_count  \
count  2949.000000  2949.000000  2949.000000  2949.000000  2949.000000   
mean     14.730417     1.944049    51.005697    76.846389    99.960326   
std       7.270474     1.433010    30.944385    23.619650     9.553398   
min       0.000000     0.000000     6.400000    33.000000    87.000000   
25%       6.000000     0.000000     8.800000    44.000000    87.000000   
50%      20.000000     3.000000    73.800000    94.000000   107.000000   
75%      20.000000     3.000000    73.800000    94.000000   107.000000   
max      20.000000     3.000000    73.800000    94.000000   107.000000   

       image_upload_id  
count              0.0  
mean               NaN  
std                NaN  
min                NaN  
25%                NaN  
50%                NaN  
75%                NaN  
max                NaN

Se puede observar que se tienen escalas diferentes por lo que posiblemente se requiera una estandarización de la información.

# Data cleaning

Se elima la columna 'imagen_upload_id' ya que de acuerdo a la consulta de los campos, no contiene información'.

In [42]:
data.drop(['image_upload_id'], axis='columns', inplace=True)
data.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2949 entries, 0 to 2948
Columns: 23 entries, username to archetype
dtypes: float64(1), int64(12), object(10)
memory usage: 530.0+ KB


In [43]:
# Verificar registros duplicados
data.loc[data.duplicated()]

Empty DataFrame
Columns: [username, trust_level, topics_entered, time_read, days_visited, posts_read_count, likes_given, likes_received, new_since, post_count, topic_count, user_custom_field, like_count, score, reads, word_count, topic, category, category_description, tag, given_date, post, archetype]
Index: []

[0 rows x 23 columns]

No existen datos duplicados en el dataset.

Ahora, se reemplazarán los valores nulos de la columna '' con 'No aplica'; ya que al eliminarlos podría perderse información relevante de las otras características. Adicional, es posible decir que el post realizado no contiene una categorías asignada.

In [44]:
# Reemplazar los valores nulos con 'No aplica'
data['category_description'] = data['category_description'].fillna('No aplica')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2949 entries, 0 to 2948
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   username              2949 non-null   object 
 1   trust_level           2949 non-null   int64  
 2   topics_entered        2949 non-null   int64  
 3   time_read             2949 non-null   int64  
 4   days_visited          2949 non-null   int64  
 5   posts_read_count      2949 non-null   int64  
 6   likes_given           2949 non-null   int64  
 7   likes_received        2949 non-null   int64  
 8   new_since             2949 non-null   object 
 9   post_count            2949 non-null   int64  
 10  topic_count           2949 non-null   int64  
 11  user_custom_field     2949 non-null   object 
 12  like_count            2949 non-null   int64  
 13  score                 2949 non-null   float64
 14  reads                 2949 non-null   int64  
 15  word_count           